### PR TITLE
Support proxies with authentication

### PR DESCRIPTION
**Pull request by jianghaolu**

### CODE REVIEW

**Comment by jianghaolu** [comment]
This will enable users to make one call to enable proxy authentication, instead of copying an Oracle sample into their code base.

**Comment by henrik-me** [comment]
@jianghaolu : thanks for your contribution. We will be looking at this and provide our feedback.

File: src/main/java/com/microsoft/aad/adal4j/SSLProxyTunnelSocketFactory.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    public Socket createSocket(InetAddress host, int port) throws IOException {
+        return createSocket(null, host.getHostName(), port, true);
+    }
+
+    public Socket createSocket(InetAddress address, int port,
+                               InetAddress clientAddress, int clientPort) throws IOException {
+        return createSocket(null, address.getHostName(), port, true);
+    }
+
+    public Socket createSocket(Socket s, String host, int port,
```
Comment (sangonzal) [inline src/main/java/com/microsoft/aad/adal4j/SSLProxyTunnelSocketFactory.java]: Socket s is passed in but never used. Perhaps if s is not null, tunnel should be set to s ?

**Comment by jianghaolu** [comment]
@SomkaPe @sangonzal 
Thank you for the review! We've submitted this usage to the legal team and they have approved this usage on OSS portal: https://ossmsft.visualstudio.com/DefaultCollection/Reviews/_workitems/edit/22799. 

I'm able to test it end-to-end but not able to add a unit test. The closest I can find is in `HttpHelperTest` but there's no example of testing a unauthenticated proxy either. Please advise.

I'm actually modifying this slightly to support challenge based authentication from the proxy server. We have customers behind a proxy that requires a digest authentication. If it's okay with you, I plan to push these changes, together with the 3rd party notice for legal purposes, in the next couple days and ask you to reconsider it.

**Comment by jianghaolu** [comment]
Re-opened here: https://github.com/AzureAD/azure-activedirectory-library-for-java/pull/255, please reconsider. Thanks!

**Pull request closed without merging**